File: intern.go
package main

import "fmt"

type Intern struct {
	fwd []uint
}

func NewIntern() *Intern {
	return &Intern{
		fwd: []uint{0},
	}
}

func (i *Intern) Add(s string) uint {
	var index uint = 0

	for j := 0; j < len(s); j++ {
		c := s[j]

		block := i.fwd[index]
		if block == 0 {
			block = uint(len(i.fwd))

			newBlock := make([]uint, 256)
			i.fwd = append(i.fwd, newBlock...)
			i.fwd[index] = block
		}
		index = uint(block) + uint(c)
	}
	return index
}

func main() {
	intern := NewIntern()

	id1 := intern.Add("hello")
	id2 := intern.Add("world")
	id3 := intern.Add("hello")

	fmt.Printf("ID 'hello': %d\n", id1)
	fmt.Printf("ID 'world': %d\n", id2)
	fmt.Printf("ID 'hello': %d\n", id3)
}

File: intern.cpp
#include <iostream>
 
class Intern {
public:
  uint add(const std::string& str) {
    uint index = -1;
    uint block = 0;
    char c;
    for (uint c : str) {
      if (block >= fwd.size()) {
        block = fwd.size();
        fwd.resize(block + 256, -1);
        bwd.push_back(index);
        if (index != (uint) -1)
          fwd[index] = block;
      }
      index = block + c;
      block = fwd[index];
    }
    return index;
  }

  std::string get(uint index) {
    std::string str;
    while (index != (unsigned)-1) {
      str.insert(0, 1, index % 256);
      index = bwd[index / 256];
    }
    return str;
  }

  private:
    std::vector<uint> fwd;
    std::vector<uint> bwd;
};
 
int main()
{
  Intern intern;

  uint id = intern.add("ABC");
  uint id2 = intern.add("DEF");
  std::cout << id << std::endl;
  std::cout << intern.get(id) << std::endl;
  std::cout << id2 << std::endl;
  std::cout << intern.get(id2) << std::endl;

  return 0;
}